### PR TITLE
Remove bind-to-device support where unavailable

### PR DESCRIPTION
**Pull request by EriComic**
On freeBSD-like network stacks, the SO_BINDTODEVICE socket option is not available. Since there is no suitable replacement for this socket option, we ignore the 'device' configuration option on systems missing the option and emit a warning instead.

This makes vsomeip compatible with the new io-sock network stack on QNX 7.1+.

Co-author: Daniel Freiermuth danielfr@haleytek.com @danielfr-haleytek

### CODE REVIEW

**Comment by duartenfonseca** [comment]
@EriComic does this mean that without the SO_BINDTODEVICE, you aren't able to bind and connect to another device?

**Comment by EriComic** [comment]
@duartenfonseca Specifically it is to bind to the _interface_ instead of just binding to the IP like we normally do. After a lot of exhaustive research in trying to find a freeBSD alternative, we found that there sadly isn't a proper replacement for this. The functionality could be brought back but the implementation would be very use-case specific and would not belong here. It is worth noting is that [QNX's port of vsomeip](https://gitlab.com/qnx/ports/vsomeip/-/blob/QNX_7.1_v3.1.20.3/implementation/endpoints/src/tcp_server_endpoint_impl.cpp#L69) also deprecates the feature.

So we moved on to evaluating if the functionality is truly needed. As mentioned, this functionality allows binding to the _interface_ instead of just the _IP_. This allows for things like:
* Filtering traffic to a single interface in multi-interface systems where the interfaces share the same IP
* Listen to traffic from an interface instead in cases where IPs have been changed/don't have an IP

These use cases are so specialized that traditional IP binding wouldn't work. Therefore, we decided to deprecate the feature and leave re-implementation of it as an option for any party that needs it.

TL;DR No, this won't prevent binding and connecting to another device. The option for _interface_ binding is removed which won't affect the more common IP binding. Interface binding is a very niche feature which we decided to drop in favor of classic IP binding.

File: implementation/endpoints/src/tcp_client_endpoint_impl.cpp [hunk 1]
```diff
@@ -167,10 +167,15 @@ void tcp_client_endpoint_impl::connect() {
         // If specified, bind to device
         std::string its_device(configuration_->get_device());
         if (its_device != "") {
+#if defined(SO_BINDTODEVICE)
             if (setsockopt(socket_->native_handle(),
                     SOL_SOCKET, SO_BINDTODEVICE, its_device.c_str(), static_cast<socklen_t>(its_device.size())) == -1) {
                 VSOMEIP_WARNING << "TCP Client: Could not bind to device \"" << its_device << "\"";
             }
+#else
+            VSOMEIP_WARNING << "TCP Client: Could not bind to device \"" << its_device << "\""
```
Comment (goncaloalmeida) [inline implementation/endpoints/src/tcp_client_endpoint_impl.cpp:176]: All the changes seem ok, but why not use the normal logging? VSOMEIP_WARNING << "tcp_client_endpoint::connect:.."  or  VSOMEIP_WARNING <<"tcp_client_endpoint::" << __func__ ""..

Comment (EriComic) [reply implementation/endpoints/src/tcp_client_endpoint_impl.cpp:176]: I see your point and I had a few ideas on how to make this better as well, but I decided against it for the following reasons:

1. The warning log that was already present in the code uses the same formatting I have currently, I believe it makes sense to adhere to what's already there. I believe best practice is to keep logging style/formatting changes in separate commits in order to maintain a "prettier git log".
2. Having function signatures in logs is very useful for developers working on the library but it's too noisy for the user using the library. Your proposed log would make more sense for a DEBUG or TRACE level log. Perhaps in addition to the currently proposed WARNING log.

TL;DR I think your suggestion is good but only useful for the _developer_ and not-so-much the _user_. If you would like the addition of this  log as a DEBUG or TRACE, then please add it, but I believe our proposed WARNING should be there as it is.
